### ISSUE TITLE
Get-Help does not return dynamic parameters or syntax

### ISSUE THREAD

**Issue by ThomasNieto** [PowerShell/PowerShell#19301]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

`Get-Help -Parameter` does not return dynamic parameter help or syntax. Related provider `-Path` issue: #19300 

The following example will return different results as described in https://github.com/PowerShell/PowerShell/issues/19301#issuecomment-1463997962 if `Update-Help` has been ran on the node.

You can see here that `Get-Command` is able to return the synatx as well as dynamic parameters for a given command. `[-Type <RegistryValueKind>]` is a dynamic parameter for the registry provider. This is due to a `ArgumentList` parameter on `Get-Commmand`. A similar parameter should be added to `Get-Help`.

The help information can be pulled from `[Parameter(HelpMessage="Descriptive parmaeter help")]` if there is no accompanying documentation in a help file. The `RuntimeDefinedParameterDictionary` class also has a `HelpFile` property that could be leveraged as well if the help is in a different module. There are currently no code paths that reference that property at the moment.

```powershell
Get-Command Set-Item -Syntax -path hklm:\

Set-Item [-Path] <string[]> [[-Value] <Object>] [-Force] [-PassThru] [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Type <RegistryValueKind>] [<CommonParameters>]

Set-Item [[-Value] <Object>] -LiteralPath <string[]> [-Force] [-PassThru] [-Filter <string>] [-Include <string[]>] [-Exclude <string[]>] [-Credential <pscredential>] [-WhatIf] [-Confirm] [-Type <RegistryValueKind>] [<CommonParameters>]
```

### Expected behavior

```console
PS C:\> Get-Help Set-Item -Path HKLM: -Parameter Type

# Parameter help is returned
```

### Actual behavior

```console
PS C:\> Get-Help Set-Item -Path HKLM: -Parameter Type
Get-Help: No parameter matches criteria Type.
```

### Error details

```console
Exception             :
    Type        : System.Management.Automation.PSArgumentException
    ErrorRecord :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : No parameter matches criteria Type.
            HResult : -2146233087
        CategoryInfo          : InvalidArgument: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : Argument
    Message     : No parameter matches criteria Type.
    ParamName   : Parameter
    HResult     : -2147024809
TargetObject          : System.Management.Automation.ProviderCommandHelpInfo
CategoryInfo          : InvalidArgument: (System.Management.A…iderCommandHelpInfo:ProviderCommandHelpInfo) [Get-Help],
PSArgumentException
FullyQualifiedErrorId : NoParmsFound,Microsoft.PowerShell.Commands.GetHelpCommand
InvocationInfo        :
    MyCommand        : Get-Help
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 16
    Line             : Get-Help Set-Item -Path HKLM: -Parameter Type
    PositionMessage  : At line:1 char:1
                       + Get-Help Set-Item -Path HKLM: -Parameter Type
                       + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    InvocationName   : Get-Help
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo :
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0-preview.1
PSEdition                      Core
GitCommitId                    7.4.0-preview.1
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
Hmm... I actually do get the expected output for `Get-Help Set-Item -Path HKLM: -Parameter Type` and it even works with `Get-Help Set-Item -Parameter Type` (_without_ a `-Path` argument), in both 7.3.3 and 7.4.0-preview.1

Not sure if it makes a difference, but have you tried running `Update-Help -Force`?

**Comment by mklement0** [comment]
The same _may_ apply analogously to #19300: perhaps missing provider-specific help files are the problem, not `Get-Help` itself; from [the `Get-Help` docs re `-Path`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/get-help):

> This parameter is effective only for help about a provider cmdlet and only when the provider includes a custom version of the provider cmdlet help article in its help file. To use this parameter, install the help file for the module that includes the provider.

**Comment by ThomasNieto** [comment]
@mklement0 the reason it works for type is because its explicitly defined in the [Set-Item](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.management/set-item?view=powershell-7.3) however I doubt it would return results lets say if `SqlServer` module had a dynamic parameter for `Set-Item`.

PowerShell help by MS has moved away from using the dedicated provider and dynamic parameter help sections in the MAML most likely due to PlatyPS and the lack of support for provider sections. I used Type in the example since it was also one that also exists on PS v2. The same issue also exists on PS v2 as that one doesn't have updateable help to invalidate the test with the missing provider MAML.

**Comment by mklement0** [comment]
That makes sense - perhaps you can update the initial post to clarify that not everyone will see the symptom, depending on whether they've run `Update-Help`.

**Comment by ThomasNieto** [comment]
@mklement0 will do.

I think I found why it doesn't work. The dynamic parameter help code is in a `if V2` section.

https://github.com/PowerShell/PowerShell/blob/b58db8b8f70756301d63ca9c52d0cca69756ef73/src/System.Management.Automation/help/MamlCommandHelpInfo.cs#L273-L285

**Comment by mklement0** [comment]
Nice sleuthing.

As for:

> This is due to a `ArgumentList` parameter on `Get-Command`. A similar parameter should be added to `Get-Help`.

`Get-Help`'s _analogous_ parameter is `-Path`, which is _non-positional_, unlike `Get-Command`'s `-ArgumentList` (`-Args`), which is a `ValueFromRemainingArguments` parameter.

If I understand correctly, the (currently broken for `Get-Help`) purpose of both these parameters is to specify a _provider context_, via a _provider-specific path_ (a _drive spec_ is enough), in order to make that provider's dynamic parameters discoverable.

As such, the following commands should display the registry provider's `-Type` parameter in their syntax diagrams (but, due to the bug you've discovered, for `Get-Help` it doesn't work / only happens to work if the help topic happens to cover _all_ providers):

```powershell
Get-Help Set-Item -Path hkcu:
Get-Command Set-Item -Path hkcu: -Syntax
```

Because of  `Get-Command`'s `ValueFromRemainingArguments` property of `-ArgumentList`, `-Path hkcu:` is actually equivalent to `-ArgumentList '-Path', 'hkcu:'`, and you can _omit_ `-Path`: 
`Get-Command Set-Item hkcu: -Syntax`

This does _not_ work with `Get-Help`, whose `-Path` parameter is non-positional.

The ability to _omit_ `-Path` with `Get-Command` is its only advantage, and implementing the same behavior for `Get-Help` would make sense.

Conversely, the `-ArgumentList` parameter with its ability to accept _multiple_ pass-through arguments is currently overkill, given that only _one_ argument is seemingly ever paid attention to, namely one that is a _provider path / drive spec_.

However, if pass-through arguments are parameter _names_, their existence on the target command _is_ enforced.

The only reason to support _multiple_ arguments would be to honor attempts to specify a provider context _via one of that provider's dynamic parameters_ as well:

```powershell
# Currently does NOT work - unless the current location already happens to be a registry location, 
# in which case you don't need the -Args argument at all.
Get-Command -Syntax Set-Item -Args '-Type', 'Dword'
```

The question, however, is whether that is worth doing, given that you then also have to specify a valid _value_ for the dynamic parameter, and that you by definition already know the name of the _specified_ dynamic parameter - it would only help with discovering _additional_ ones.

---

In short: a simple solution that doesn't introduce new capabilities and focuses on harmonizing the syntax between the two cmdlets would be the following:

* Make `Get-Help`'s `-Path` parameter positional (there is no backward-compatibility concern, given that there's currently only _one_ positional parameter, `-Name`).

* Introduce `-Path` as an alias for `Get-Command`'s `-ArgumentList` parameter and clearly document that it only ever makes sense to pass _one_ argument, namely a provider path / drive spec.

**Comment by ThomasNieto** [comment]
> If I understand correctly, the (currently broken for Get-Help) purpose of both these parameters is to specify a provider context, via a provider-specific path (a drive spec is enough), in order to make that provider's dynamic parameters discoverable.

While `-Path` and `-ArgumentList` are both defined to accomplish similar results there is a distinction. `Get-Help -Path` is intended to return contextual provider help be that parameter help, cmdlet description, etc. Meanwhile `Get-Command -ArgumentList` is not limited to just providers but can be used any cmdlet that implements dynamic parameters.

> Conversely, the -ArgumentList parameter with its ability to accept multiple pass-through arguments is currently overkill, given that only one argument is seemingly ever paid attention to, namely one that is a provider path / drive spec.

While it seems like overkill and in most cases is, it is required because any cmdlet that implements dynamic parameters can use single or multiple parameters as keys to change the command syntax. (Using multiple parameters as context to generate dynamic parameters would be bad user experience).

> Introduce -Path as an alias for Get-Command's -ArgumentList parameter and clearly document that it only ever makes sense to pass one argument, namely a provider path / drive spec.

I believe this would break cmdlets that implement dynamic parameters and use `-Path` parameter key. This would swallow the named parameter and pass it as a positional one.

## 

This issue stems from my desire to add help for my module's dynamic parameters. There is a bug and a request in this issue. I can create a separate issue for my request if that would help.

1. Provider based dynamic parameters is broken due to the bug identified here
2. Get-Help should work with any cmdlet that has dynamic parameters if that help is defined in the current module or another

I am not sure if the bug was fixed and [Provider dynamic parameter help](https://learn.microsoft.com/en-us/powershell/scripting/developer/help/how-to-add-dynamic-parameters-to-a-provider-help-topic) document was followed if it would enable any cmdlet with dynamic parameter help.

Here is an example from my module that is similar to `PSProviders` with cmdlets that are a framework that call other another module's implementation and can define dynamic parameters.

```powershell
Install-PSResource AnyPackage.Scoop
Import-Module AnyPackage.Scoop

Get-Command Register-PackageSource -Provider Scoop -Syntax

Register-PackageSource [-Name] <string> [-Location] <string> [-Provider] <string> [-Trusted] [-PassThru] [-Force] [-WhatIf] [-Confirm] [<CommonParameters>]

Register-PackageSource [-Provider] <string> -Official <string> [-Trusted] [-PassThru] [-Force] [-WhatIf] [-Confirm] [<CommonParameters>]

Get-Help Register-PackageSource -Provider Scoop
```

The cmdlet `Register-PackageSource` is shipped in the module `AnyPackage`. The parameter `-Official` is a dynamic parameter when `-Provider` is passed and the value is `Scoop` from the `AnyPacakge.Scoop` module. The PSProvider system does this with `-Path` behind the scenes since it uses `$PWD` as the default value.

The `Get-Help` command in the example would fail since the cmdlet is not a provider and doesn't use `-Path` as a key. This is why I mention adding a `-ArgumentList` parameter to `Get-Help` to allow for non-provider cmdlets to return dynamic parameter help.

**Comment by mklement0** [comment]
All makes sense; thanks for the detailed explanation - I've hidden my previous comment.

I hadn't considered the case where the availability of dynamic parameters can depend on also specifying arbitrary _other_ parameters for a given cmdlet, not just depend on the provider context (if any).

**Comment by sdwheeler** [comment]
The biggest issue we have is that the tooling of our publishing pipeline doesn't support this documentation scenario. This is something we want to fix at some point, but there are a lot of dependencies that need to be delivered first.
